# Project 3 -- Lukas Anell

In [2]:
# /anvil/projects/tdm/data/grouping/vgchartz-2024.csv
# /anvil/projects/tdm/data/grouping/googleplaystore.csv
# /anvil/projects/tdm/data/election/itcont1980.txt (used in the video)

## Question 1

In [2]:
library(data.table)
video_games <- as.data.frame(fread("/anvil/projects/tdm/data/grouping/vgchartz-2024.csv"))
head(video_games)

img                                                                 
1 /games/boxart/full_6510540AmericaFrontccc.jpg                       
2 /games/boxart/full_5563178AmericaFrontccc.jpg                       
3 /games/boxart/827563ccc.jpg                                         
4 /games/boxart/full_9218923AmericaFrontccc.jpg                       
5 /games/boxart/full_4990510AmericaFrontccc.jpg                       
6 /games/boxart/full_call-of-duty-modern-warfare-3_517AmericaFront.jpg
  title                          console genre   publisher      developer     
1 Grand Theft Auto V             PS3     Action  Rockstar Games Rockstar North
2 Grand Theft Auto V             PS4     Action  Rockstar Games Rockstar North
3 Grand Theft Auto: Vice City    PS2     Action  Rockstar Games Rockstar North
4 Grand Theft Auto V             X360    Action  Rockstar Games Rockstar North
5 Call of Duty: Black Ops 3      PS4     Shooter Activision     Treyarch      
6 Call of Duty: Modern Warfare 3 X360    Shooter Activision     Infinity Ward 
  critic_score total_sales na_sales jp_sales pal_sales other_sales release_date
1 9.4          20.32       6.37     0.99     9.85      3.12        2013-09-17  
2 9.7          19.39       6.06     0.60     9.71      3.02        2014-11-18  
3 9.6          16.15       8.41     0.47     5.49      1.78        2002-10-28  
4  NA          15.86       9.06     0.06     5.33      1.42        2013-09-17  
5 8.1          15.09       6.18     0.41     6.05      2.44        2015-11-06  
6 8.7          14.82       9.07     0.13     4.29      1.33        2011-11-08  
  last_update
1 <NA>       
2 2018-01-03 
3 <NA>       
4 <NA>       
5 2018-01-14 
6 <NA>

In [3]:
tapply(video_games$critic_score, video_games$genre, mean)

Action Action-Adventure        Adventure       Board Game 
              NA               NA               NA               NA 
       Education         Fighting              MMO             Misc 
              NA               NA               NA               NA 
           Music            Party         Platform           Puzzle 
              NA               NA               NA               NA 
          Racing     Role-Playing          Sandbox          Shooter 
              NA               NA               NA               NA 
      Simulation           Sports         Strategy     Visual Novel 
              NA               NA               NA               NA

In [4]:
genreCriticScores <- tapply(video_games$critic_score, video_games$genre, mean, na.rm=TRUE)
genreCriticScores

Action Action-Adventure        Adventure       Board Game 
        6.831008         7.678378         7.341058         5.300000 
       Education         Fighting              MMO             Misc 
        7.750000         7.392424         7.781818         6.884593 
           Music            Party         Platform           Puzzle 
        7.992593         5.968750         7.307638         7.201049 
          Racing     Role-Playing          Sandbox          Shooter 
        7.187524         7.470920         9.200000         7.276800 
      Simulation           Sports         Strategy     Visual Novel 
        6.914719         7.291027         7.431733         8.500000

In [5]:
totalSalesConsoles <- tapply(video_games$total_sales, video_games$console, sum, na.rm=TRUE)
totalSalesConsoles

2600     3DO     3DS    5200    7800    ACPC      AJ     AST     Aco     All 
  75.66    0.19   99.27    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   Amig     And    ApII     Arc    BBCM     BRW    C128     C64    CD32     CDi 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
     CV      DC      DS     DSi    DSiW     FDS     FMT      GB     GBA     GBC 
   0.00   13.31  458.17    0.00    0.00    0.00    0.00   19.84  224.48    4.34 
     GC     GEN      GG     GIZ     Int   Linux    Lynx      MS     MSD     MSX 
 119.53   18.83    0.04    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
    Mob     N64     NES      NG   NGage      NS      OR     OSX    Ouya      PC 
   0.03   93.79   47.93    1.44    0.00   36.46    0.00    0.03    0.00  168.99 
    PCE    PCFX      PS     PS2     PS3     PS4     PS5     PSN     PSP     PSV 
   0.16    0.03  546.25 1027.76  839.70  539.92    0.00    3.81  245.29   63.02 
   S32X     SAT     SCD    SNES  Series    TG16      VB      VC      WS      WW 
   0.00   33.67    0.37   65.71    0.00    0.00    0.00    0.35    1.43    0.25 
    Wii    WiiU    WinP    X360      XB     XBL    XOne      XS     ZXS     iOS 
 459.44   35.42    0.00  859.79  232.05    0.20  268.96    0.00    0.00    0.00 
   iQue 
   0.00

In [6]:
sort(totalSalesConsoles, decreasing = TRUE)

PS2    X360     PS3      PS     PS4     Wii      DS    XOne     PSP      XB 
1027.76  859.79  839.70  546.25  539.92  459.44  458.17  268.96  245.29  232.05 
    GBA      PC      GC     3DS     N64    2600    SNES     PSV     NES      NS 
 224.48  168.99  119.53   99.27   93.79   75.66   65.71   63.02   47.93   36.46 
   WiiU     SAT      GB     GEN      DC     GBC     PSN      NG      WS     SCD 
  35.42   33.67   19.84   18.83   13.31    4.34    3.81    1.44    1.43    0.37 
     VC      WW     XBL     3DO     PCE      GG     Mob     OSX    PCFX    5200 
   0.35    0.25    0.20    0.19    0.16    0.04    0.03    0.03    0.03    0.00 
   7800    ACPC      AJ     AST     Aco     All    Amig     And    ApII     Arc 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   BBCM     BRW    C128     C64    CD32     CDi      CV     DSi    DSiW     FDS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
    FMT     GIZ     Int   Linux    Lynx      MS     MSD     MSX   NGage      OR 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   Ouya     PS5    S32X  Series    TG16      VB    WinP      XS     ZXS     iOS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   iQue 
   0.00

In [7]:
sort(genreCriticScores, decreasing = TRUE)

Sandbox     Visual Novel            Music              MMO 
        9.200000         8.500000         7.992593         7.781818 
       Education Action-Adventure     Role-Playing         Strategy 
        7.750000         7.678378         7.470920         7.431733 
        Fighting        Adventure         Platform           Sports 
        7.392424         7.341058         7.307638         7.291027 
         Shooter           Puzzle           Racing       Simulation 
        7.276800         7.201049         7.187524         6.914719 
            Misc           Action            Party       Board Game 
        6.884593         6.831008         5.968750         5.300000

In [8]:
sort(totalSalesConsoles, decreasing = TRUE)

PS2    X360     PS3      PS     PS4     Wii      DS    XOne     PSP      XB 
1027.76  859.79  839.70  546.25  539.92  459.44  458.17  268.96  245.29  232.05 
    GBA      PC      GC     3DS     N64    2600    SNES     PSV     NES      NS 
 224.48  168.99  119.53   99.27   93.79   75.66   65.71   63.02   47.93   36.46 
   WiiU     SAT      GB     GEN      DC     GBC     PSN      NG      WS     SCD 
  35.42   33.67   19.84   18.83   13.31    4.34    3.81    1.44    1.43    0.37 
     VC      WW     XBL     3DO     PCE      GG     Mob     OSX    PCFX    5200 
   0.35    0.25    0.20    0.19    0.16    0.04    0.03    0.03    0.03    0.00 
   7800    ACPC      AJ     AST     Aco     All    Amig     And    ApII     Arc 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   BBCM     BRW    C128     C64    CD32     CDi      CV     DSi    DSiW     FDS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
    FMT     GIZ     Int   Linux    Lynx      MS     MSD     MSX   NGage      OR 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   Ouya     PS5    S32X  Series    TG16      VB    WinP      XS     ZXS     iOS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   iQue 
   0.00

In [9]:
sort(tapply(video_games$title, video_games$genre, FUN = length), decreasing = TRUE)

Misc           Action        Adventure     Role-Playing 
            9304             8557             6260             5721 
          Sports          Shooter         Platform         Strategy 
            5586             5410             4001             3685 
          Puzzle           Racing       Simulation         Fighting 
            3521             3425             3158             2367 
Action-Adventure     Visual Novel            Music            Party 
            1877              493              297              151 
             MMO        Education       Board Game          Sandbox 
             115               35               33               20

tapply seems like a very useful function, but it took me a bit to figure out what the first two arguments do. From what I gathered, it seems like you apply the function to one vector, then display it in terms of the other one.

na.rm seems like a very nice and simple method of removing the N/A scores by critics in this application, and I think would definitely be useful in other places.

When using sort, I'm curious if the function will always automatically sort by numbers, or even if it supports alphabetical or other kinds of sorting if numbers aren't available. Ah, reading into the instructions further, I see that you can specify the kind of sorting.

It seems like the Misc. category has the most games in it, which makes sense because it probably encompasses the most. Action does get very close, though.

1.1: Sandbox

1.2: PS2

1.3: Based on my usage, the `tapply()` function takes 3 arguments. The first one specifies the column we will be applying a function to. The second one is the column the data is grouped under. And the third one specifies the function being applied to the first argument. For example, in the title and genre step, `tapply()` will apply the `length()` function to the video game's title, and add to a total that is associated with the genre of the game.

## Question 2

In [10]:
library(dplyr)
video_games %>%
    group_by(genre) %>%
    summarize(mean_critic_score = mean(critic_score, na.rm = TRUE))


Attaching package: 'dplyr'


The following objects are masked from 'package:data.table':

    between, first, last


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




genre            mean_critic_score
1  Action           6.831008         
2  Action-Adventure 7.678378         
3  Adventure        7.341058         
4  Board Game       5.300000         
5  Education        7.750000         
6  Fighting         7.392424         
7  MMO              7.781818         
8  Misc             6.884593         
9  Music            7.992593         
10 Party            5.968750         
11 Platform         7.307638         
12 Puzzle           7.201049         
13 Racing           7.187524         
14 Role-Playing     7.470920         
15 Sandbox          9.200000         
16 Shooter          7.276800         
17 Simulation       6.914719         
18 Sports           7.291027         
19 Strategy         7.431733         
20 Visual Novel     8.500000

This new summarize function is very confusing. It feels very complex for the task, but I'm assuming it is very good with more difficult things.

2.1: Line 12

2.2: The biggest difference I noticed is that `tapply()` doesn't return a formatted table, just a list of what seems like key:value pairs? The data is presented in a less pretty way. I'd say that after reading some documentation for the `dplyr` functions, it's much easier to apply to the data frame because it's spread out across multiple lines by default. I also think it's much easier to tell how the data will be grouped and to apply functions to it.

2.3: I think I prefer `dplyr`. As I said above, I think that after reading the documentation for it, it was much easier to handle the data with just a few short lines of logic. This is in contrast to `tapply()`, where in my opinion it felt like I had to think about how the data would be presented and I had to do a lot of function wrapping instead of just adding a new line.

## Question 3

In [11]:
console_sales <- video_games %>%
    group_by(console) %>%
    summarize(sum_sales = sum(total_sales, na.rm = TRUE))
head(console_sales)

console sum_sales
1 2600    75.66    
2 3DO      0.19    
3 3DS     99.27    
4 5200     0.00    
5 7800     0.00    
6 ACPC     0.00

In [12]:
console_sales <- video_games %>%
    group_by(console) %>%
    summarize(total_console_sales = sum(total_sales, na.rm = TRUE)) %>%
    arrange(desc(total_console_sales))
head(console_sales)

console total_console_sales
1 PS2     1027.76            
2 X360     859.79            
3 PS3      839.70            
4 PS       546.25            
5 PS4      539.92            
6 Wii      459.44

In [13]:
genre_title_counts <- video_games %>%
    group_by(genre) %>%
    summarize(total_titles = n()) %>%
    arrange(desc(total_titles))
tail(genre_title_counts)

genre      total_titles
1 Music      297         
2 Party      151         
3 MMO        115         
4 Education   35         
5 Board Game  33         
6 Sandbox     20

In [14]:
publisher_total_sales_counts <- video_games %>%
    group_by(publisher) %>%
    summarize(sum_sales = sum(total_sales, na.rm = TRUE)) %>%
    arrange(desc(sum_sales))
head(publisher_total_sales_counts)

publisher                   sum_sales
1 Activision                  722.77   
2 Electronic Arts             644.13   
3 EA Sports                   485.66   
4 Ubisoft                     462.57   
5 THQ                         320.89   
6 Sony Computer Entertainment 311.08

I had to play around with `arrange()` for a little bit, as I was confused about if I had to include the other code or if it would be peristent from the line above, but I got it working eventually. It seems very nice as well, but I think my biggest confusion so far is how using these function invocations one after another works sometimes. I think I'm understanding more as I do more though.

I like `n()` a lot because it seems like it would be very versatile. I was a little confused about how to use it at first, but the documentation helped. But I'm curious, how does it automatically know what we're sorting by? I'd be interested to see how it works under the hood.

3.1: PS2

3.2: Sandbox

3.3: I decieded to show a grouping consisting of publishers and each of their respective total sales. I'm actually not too sure what the total_sales column represents, it could possibly be copies sold in some multiple, but I think it's interesting to see what publishers have sold the most games in general.

## Question 4

In [15]:
app_store <- read.csv("/anvil/projects/tdm/data/grouping/googleplaystore.csv")
head(app_store)

App                                                           Category      
1 Photo Editor & Candy Camera & Grid & ScrapBook                ART_AND_DESIGN
2 Coloring book moana                                           ART_AND_DESIGN
3 U Launcher Lite \342\200\223 FREE Live Cool Themes, Hide Apps ART_AND_DESIGN
4 Sketch - Draw & Paint                                         ART_AND_DESIGN
5 Pixel Draw - Number Art Coloring Book                         ART_AND_DESIGN
6 Paper flowers instructions                                    ART_AND_DESIGN
  Rating Reviews Size Installs    Type Price Content.Rating
1 4.1    159     19M  10,000+     Free 0     Everyone      
2 3.9    967     14M  500,000+    Free 0     Everyone      
3 4.7    87510   8.7M 5,000,000+  Free 0     Everyone      
4 4.5    215644  25M  50,000,000+ Free 0     Teen          
5 4.3    967     2.8M 100,000+    Free 0     Everyone      
6 4.4    167     5.6M 50,000+     Free 0     Everyone      
  Genres                    Last.Updated     Current.Ver        Android.Ver 
1 Art & Design              January 7, 2018  1.0.0              4.0.3 and up
2 Art & Design;Pretend Play January 15, 2018 2.0.0              4.0.3 and up
3 Art & Design              August 1, 2018   1.2.4              4.0.3 and up
4 Art & Design              June 8, 2018     Varies with device 4.2 and up  
5 Art & Design;Creativity   June 20, 2018    1.1                4.4 and up  
6 Art & Design              March 26, 2017   1.0                2.3 and up

In [16]:
app_store$Clean_Reviews <- gsub(",","",app_store$Review)
app_store$Numeric_Reviews <- as.numeric(app_store$Clean_Reviews)
app_store[is.na(app_store$Numeric_Reviews) & !is.na(app_store$Reviews), "Reviews"]

Warning message:
"NAs introduced by coercion"


[1] "3.0M"

In [49]:
app_store_cleaned <- app_store %>%
    mutate(
        Clean_Reviews = gsub(",","", Reviews),
        Numeric_Reviews = as.numeric(Clean_Reviews)
    ) %>%
    filter(!is.na(Numeric_Reviews))
head(app_store_cleaned)

Warning message:
"There was 1 warning in `mutate()`.
i In argument: `Numeric_Reviews = as.numeric(Clean_Reviews)`.
Caused by warning:
! NAs introduced by coercion"


App                                                           Category      
1 Photo Editor & Candy Camera & Grid & ScrapBook                ART_AND_DESIGN
2 Coloring book moana                                           ART_AND_DESIGN
3 U Launcher Lite \342\200\223 FREE Live Cool Themes, Hide Apps ART_AND_DESIGN
4 Sketch - Draw & Paint                                         ART_AND_DESIGN
5 Pixel Draw - Number Art Coloring Book                         ART_AND_DESIGN
6 Paper flowers instructions                                    ART_AND_DESIGN
  Rating Reviews Size Installs    Type Price Content.Rating
1 4.1    159     19M  10,000+     Free 0     Everyone      
2 3.9    967     14M  500,000+    Free 0     Everyone      
3 4.7    87510   8.7M 5,000,000+  Free 0     Everyone      
4 4.5    215644  25M  50,000,000+ Free 0     Teen          
5 4.3    967     2.8M 100,000+    Free 0     Everyone      
6 4.4    167     5.6M 50,000+     Free 0     Everyone      
  Genres                    Last.Updated     Current.Ver        Android.Ver 
1 Art & Design              January 7, 2018  1.0.0              4.0.3 and up
2 Art & Design;Pretend Play January 15, 2018 2.0.0              4.0.3 and up
3 Art & Design              August 1, 2018   1.2.4              4.0.3 and up
4 Art & Design              June 8, 2018     Varies with device 4.2 and up  
5 Art & Design;Creativity   June 20, 2018    1.1                4.4 and up  
6 Art & Design              March 26, 2017   1.0                2.3 and up  
  Clean_Reviews Numeric_Reviews
1 159              159         
2 967              967         
3 87510          87510         
4 215644        215644         
5 967              967         
6 167              167

In [31]:
app_store[is.na(app_store_cleaned$Numeric_Reviews) & !is.na(app_store_cleaned$Reviews), "Reviews"]

character(0)

In [34]:
app_store_cleaned <- app_store_cleaned %>%
    mutate(
        rating_per_review = Rating / Numeric_Reviews
    )
head(app_store_cleaned$rating_per_review)

[1] 2.578616e-02 4.033092e-03 5.370815e-05 2.086773e-05 4.446743e-03
[6] 2.634731e-02

At the time of writing this I'm a bit confused about how the `app_store[is.na(app_store$Numeric_Reviews) & !is.na(app_store$Reviews), "Reviews"]` works, because it's not printing anything out for me. Maybe that's how it's supposed to be now, I'm not sure. Nevermind, I fixed it now and I understand that it's printing out the problems.

I also think `mutate()` is a lot nicer for this type of thing, I like how each line is a new thing applied to the data table instead of having to access it again.

I'm very confused about the instructions at the very end of this question, so I'll have to keep messing around with filter to get the result it wants.

I finally figured out the problem with my `filter()`. It took me a while to realize that I was applying it correctly, but I wasn't checking app_store_cleaned as I should have, I was still checking app_store.

I like how easy it is to add columns to the data frames with the new syntax with `%>%`, it's very simple.



## Question 5

In [35]:
unique(app_store_cleaned$Content.Rating)

[1] "Everyone"        "Teen"            "Everyone 10+"    "Mature 17+"     
[5] "Adults only 18+" "Unrated"         ""

In [36]:
length(unique(app_store_cleaned$Content.Rating))

[1] 7

In [51]:
app_store_cleaned$Content.Rating_uppercase <- lapply(app_store_cleaned$Content.Rating, toupper)
head(app_store_cleaned$Content.Rating_uppercase)

[[1]]
[1] "EVERYONE"

[[2]]
[1] "EVERYONE"

[[3]]
[1] "EVERYONE"

[[4]]
[1] "TEEN"

[[5]]
[1] "EVERYONE"

[[6]]
[1] "EVERYONE"

In [47]:
class(app_store_cleaned$Content.Rating)
class(app_store_cleaned$Price)
class(app_store_cleaned$Rating)

[1] "character"

[1] "character"

[1] "numeric"

In [53]:
lapply(app_store_cleaned, class)

$App
[1] "character"

$Category
[1] "character"

$Rating
[1] "numeric"

$Reviews
[1] "character"

$Size
[1] "character"

$Installs
[1] "character"

$Type
[1] "character"

$Price
[1] "character"

$Content.Rating
[1] "character"

$Genres
[1] "character"

$Last.Updated
[1] "character"

$Current.Ver
[1] "character"

$Android.Ver
[1] "character"

$Clean_Reviews
[1] "character"

$Numeric_Reviews
[1] "numeric"

$Content.Rating_uppercase
[1] "list"

In [54]:
sapply(app_store_cleaned, class)

App                 Category                   Rating 
             "character"              "character"                "numeric" 
                 Reviews                     Size                 Installs 
             "character"              "character"              "character" 
                    Type                    Price           Content.Rating 
             "character"              "character"              "character" 
                  Genres             Last.Updated              Current.Ver 
             "character"              "character"              "character" 
             Android.Ver            Clean_Reviews          Numeric_Reviews 
             "character"              "character"                "numeric" 
Content.Rating_uppercase 
                  "list"

In [56]:
head(lapply(app_store_cleaned$Content.Rating, nchar), n=30)

[[1]]
[1] 8

[[2]]
[1] 8

[[3]]
[1] 8

[[4]]
[1] 4

[[5]]
[1] 8

[[6]]
[1] 8

[[7]]
[1] 8

[[8]]
[1] 8

[[9]]
[1] 8

[[10]]
[1] 8

[[11]]
[1] 8

[[12]]
[1] 8

[[13]]
[1] 4

[[14]]
[1] 8

[[15]]
[1] 8

[[16]]
[1] 8

[[17]]
[1] 8

[[18]]
[1] 8

[[19]]
[1] 8

[[20]]
[1] 8

[[21]]
[1] 8

[[22]]
[1] 8

[[23]]
[1] 12

[[24]]
[1] 8

[[25]]
[1] 8

[[26]]
[1] 8

[[27]]
[1] 8

[[28]]
[1] 8

[[29]]
[1] 8

[[30]]
[1] 8

In [57]:
head(sapply(app_store_cleaned$Content.Rating, nchar), n=30)

Everyone     Everyone     Everyone         Teen     Everyone     Everyone 
           8            8            8            4            8            8 
    Everyone     Everyone     Everyone     Everyone     Everyone     Everyone 
           8            8            8            8            8            8 
        Teen     Everyone     Everyone     Everyone     Everyone     Everyone 
           4            8            8            8            8            8 
    Everyone     Everyone     Everyone     Everyone Everyone 10+     Everyone 
           8            8            8            8           12            8 
    Everyone     Everyone     Everyone     Everyone     Everyone     Everyone 
           8            8            8            8            8            8

I like that there's a specific function for `unique()`, otherwise I would just assume I'd have to write it myself if I wanted to have that functionality. I guess I shouldn't be surprised though, but I do think it's very helpful.

I don't really understand when you'd use `lapply()` vs `sapply()`, they seem like almost the same thing except for how the data is returned. Is that just a preference of data presentation?

After trying the `nchar` part of this problem, I see how it would be a lot easier to understand the data with `sapply()`, but I guess `lapply()` looks a little better. I wonder if you could combine the best parts of both, like a list but it shows the **Everyone** or the rating for each entry in the list.

5.3: After doing a little bit of research, I found a plain `apply()` function. From what the [documentation](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/apply) says, it seems like this function makes it very easy to apply a function to entire rows and/or columns. I think this would be nice for changing larger portions of a dataset at once, instead of having to do a separate `lapply()` or `sapply()` for every row/column.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
